src/commands/run.rs

```diff
@@ -0,0 +1,177 @@
+                script = &self.script,
+                path = cwd.display(),
+            );
+            eprintln!("> {command}\n");
+            return run_script(&self.script, command, &cwd);
+        } else {
+            let e: Result<(), MainError> = PnError::MissingScript { name: self.script }
+                .pipe(MainError::Pn)
+                .pipe(Err);
+            return e;
+        }
+    }
+}
+
+/// Structure of `package.json`.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct NodeManifest {
+    #[serde(default)]
+    name: String,
+
+    #[serde(default)]
+    scripts: HashMap<String, String>,
+
+    #[serde(default)]
+    version: String,
+}
+
+fn read_package_manifest(manifest_path: &Path) -> Result<NodeManifest, MainError> {
+    if !manifest_path.exists() {
+        return Err(MainError::Pn(PnError::NoPkgManifest {
+            file: manifest_path.to_path_buf(),
+        }));
+    }
+
+    manifest_path
+        .pipe(File::open)
+        .map_err(MainError::from_dyn)?
+        .pipe(serde_json::de::from_reader::<_, NodeManifest>)
+        .map_err(|err| {
+            MainError::Pn(PnError::ParseJsonError {
+                file: manifest_path.to_path_buf(),
+                message: err.to_string(),
+            })
+        })
+}
+
+fn run_script(name: &str, command: &str, cwd: &Path) -> Result<(), MainError> {
+    let path_env = create_path_env()?;
+    let status = process::Command::new("sh")
+        .current_dir(cwd)
+        .env("PATH", path_env)
+        .arg("-c")
+        .arg(command)
+        .stdin(Stdio::inherit())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .spawn()
+        .map_err(MainError::from_dyn)?
+        .wait()
+        .map_err(MainError::from_dyn)?
+        .code()
+        .map(NonZeroI32::new);
+    match status {
+        Some(None) => return Ok(()),
+        Some(Some(status)) => PnError::ScriptError {
+            name: name.to_string(),
+            status,
+        },
+        None => PnError::UnexpectedTermination {
+            command: command.to_string(),
+        },
+    }
+    .pipe(MainError::Pn)
+    .pipe(Err)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pretty_assertions::assert_eq;
+
+    #[test]
+    fn test_read_package_manifest_ok() {
+        use std::fs;
+        use tempfile::tempdir;
+
+        let temp_dir = tempdir().unwrap();
+        let package_json_path = temp_dir.path().join("package.json");
+        fs::write(
+            &package_json_path,
+            r#"{"name": "", "scripts": {"test": "echo hello world"}, "version": ""}"#,
+        )
+        .unwrap();
+
+        let package_manifest = read_package_manifest(&package_json_path).unwrap();
+
+        let received = serde_json::to_string_pretty(&package_manifest).unwrap();
+        let expected = serde_json::json!({
+            "name": "",
+            "version": "",
```

Comment:
If `name` and `version` would become optional, this can be removed.

```suggestion
```

Reply:
they are optional but it does end up failing the test since converting the NodeManifest struct to json includes the empty name and version in the output.

Reply:
You can compare `serde_json::Value`: Replace `to_string_pretty` with `to_value` in the `received` expression and remove `to_string_pretty` in the `expected` expression.

Reply:
```rust
        let received = serde_json::to_value(&package_manifest).unwrap();
        let expected = serde_json::json!({
            "name": "",
            "version": "",
            "scripts": {
                "test": "echo hello world"
            }
        });
```